runtime/drivers/snowflake/sql_store.go

```diff
@@ -0,0 +1,250 @@
+func (c *connection) QueryAsFiles(ctx context.Context, props map[string]any, opt *drivers.QueryOption, p drivers.Progress) (drivers.FileIterator, error) {
+	srcProps, err := parseSourceProperties(props)
+	if err != nil {
+		return nil, err
+	}
+
+	var dsn string
+	if srcProps.DSN != "" { // get from src properties
+		dsn = srcProps.DSN
+	} else if url, ok := c.config["dsn"].(string); ok && url != "" { // get from driver configs
+		dsn = url
+	} else {
+		return nil, fmt.Errorf("the property 'dsn' is required for Snowflake. Provide 'dsn' in the YAML properties or pass '--env connector.snowflake.dsn=...' to 'rill start'")
+	}
+
+	db, err := sql.Open("snowflake", dsn)
+	if err != nil {
+		log.Fatalf("failed to connect. %v, err: %v", dsn, err)
+	}
+
+	ctx = sf.WithOriginalTimestamp(sf.WithArrowAllocator(sf.WithArrowBatches(ctx), memory.DefaultAllocator))
+
+	conn, err := db.Conn(ctx)
+	if err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	var rows sqld.Rows
+	err = conn.Raw(func(x interface{}) error {
+		rows, err = x.(sqld.QueryerContext).QueryContext(ctx, srcProps.SQL, nil)
+		return err
+	})
+	if err != nil {
+		rows.Close()
+		conn.Close()
+		db.Close()
+		log.Fatalf("unable to run the query. err: %v", err)
+	}
+
+	batches, err := rows.(sf.SnowflakeRows).GetArrowBatches()
+
+	// the number of returned rows is unknown at this point, only the number of batches and output files
+	p.Target(1, drivers.ProgressUnitFile)
+
+	return &fileIterator{
+		ctx:          ctx,
+		db:           db,
+		conn:         conn,
+		rows:         rows,
+		batches:      batches,
+		progress:     p,
+		limitInBytes: opt.TotalLimitInBytes,
+		logger:       c.logger,
+	}, nil
+}
+
+type fileIterator struct {
+	ctx          context.Context
+	db           *sql.DB
+	conn         *sql.Conn
+	rows         sqld.Rows
+	batches      []*sf.ArrowBatch
+	progress     drivers.Progress
+	limitInBytes int64
+	logger       *zap.Logger
+	// Computed while iterating
+	totalRecords int64
+	tempFilePath string
+	downloaded   bool
+}
+
+// Close implements drivers.FileIterator.
+func (f *fileIterator) Close() error {
+	return os.Remove(f.tempFilePath)
+}
+
+// Next implements drivers.FileIterator.
+// Query result is written to a single parquet file.
+func (f *fileIterator) Next() ([]string, error) {
+	// close db resources early
+	defer func() {
+		f.rows.Close()
+		f.conn.Close()
+		f.db.Close()
+	}()
+
+	if f.downloaded {
+		return nil, io.EOF
+	}
+	f.logger.Info("downloading results in parquet file", observability.ZapCtx(f.ctx))
+
+	// create a temp file
+	fw, err := os.CreateTemp("", "temp*.parquet")
+	if err != nil {
+		return nil, err
+	}
+	defer fw.Close()
+	f.tempFilePath = fw.Name()
+	f.downloaded = true
+
+	tf := time.Now()
+	defer func() {
+		f.logger.Info("time taken to write arrow records in parquet file", zap.Duration("duration", time.Since(tf)), observability.ZapCtx(f.ctx))
+	}()
+
+	if len(f.batches) == 0 {
+		// empty result
+		return nil, nil
```

Comment:
I think this should return some error like `no results found for the query`. Check postgres/bigquery.
Also this can be handled in `QueryAsFiles` itself.